osutil/dm_ioctl_linux.go

```diff
@@ -0,0 +1,121 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package osutil
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"os"
+	"syscall"
+	"unsafe"
+
+	"golang.org/x/sys/unix"
+)
+
+var (
+	osOpenFile = os.OpenFile
+)
+
+const (
+	DM_VERSION_MAJOR      = 4
+	DM_VERSION_MINOR      = 0
+	DM_VERSION_PATCHLEVEL = 0
+)
+
+func dmIoctlImpl(fd uintptr, command int, data unsafe.Pointer) error {
+	_, _, errno := syscall.Syscall(syscall.SYS_IOCTL, fd, uintptr(command), uintptr(data))
+	if errno != 0 {
+		return os.NewSyscallError("ioctl", errno)
+	}
+	return nil
+}
+
+var dmIoctl = dmIoctlImpl
+
+type TargetInfo = struct {
+	SectorStart uint64
+	Length      uint64
+	Status      int32
+	TargetType  string
+	Params      string
+}
+
+func DmIoctlTableStatus(major uint32, minor uint32) ([]TargetInfo, error) {
+	dmControl, err := osOpenFile("/dev/mapper/control", os.O_RDWR, 0)
```

Comment:
Can you link to https://elixir.bootlin.com/linux/v6.10.6/source/drivers/md/dm-ioctl.c and https://elixir.bootlin.com/linux/v6.10.6/source/include/uapi/linux/dm-ioctl.h, it seems that this may be the only useful documentation for the ioctl interface right now.

Reply:
and maybe https://gitlab.com/lvmteam/lvm2/-/blob/main/device_mapper/ioctl/libdm-iface.c  ? eh there isn't much documentation available

Reply:
Maybe just `/usr/include/linux/dm-ioctl.h`.